Rust UI: improve messages returned to python

mmilata opened the pull request:
Built on top of #2092. Pyright fails since we no longer generate the `.pyi` files, can we do better than just ignoring these issues? Some other questions inline.

Fixes #2118.

mmilata commented on core/embed/rust/src/ui/layout/result.rs:
```diff
@@ -0,0 +1,34 @@
+unsafe impl Sync for ResultObj {}
+
+static CONFIRMED_TYPE: Type = obj_type! { name: Qstr::MP_QSTR_CONFIRMED, };
+static CANCELLED_TYPE: Type = obj_type! { name: Qstr::MP_QSTR_CANCELLED, };
```
Implemented this as two distinct types because it was the least comlex way to make python `print()` the value. Is there a reason not to do it like this.

jpochyla replied:
It takes a bit more flash but I wouldn't sweat it too much.

matejcik replied:
For a smaller version, we could just make two `mp_obj_object_t`s. The option to print the object name wouldn't be there but we don't _really_ need that.

But this is good enough I think.

mmilata replied:
Ok, I'll leave it like this now and rewrite it once we add more result constants.

mmilata commented on core/embed/rust/src/ui/model_t1/layout.rs:
```diff
@@ -60,12 +58,7 @@ extern "C" fn ui_layout_new_confirm_action(
-    args: *const Obj,
-    kwargs: *const Map,
-) -> Obj {
+extern "C" fn new_confirm_text(n_args: usize, args: *const Obj, kwargs: *mut Map) -> Obj {
```
Note that I changed `*const Map` to `*mut Map` to pass type check.

jpochyla replied:
I can't check it now, but it should match the type the C-side expects.

mmilata commented on core/embed/rust/src/ui/model_tt/component/keyboard/pin.rs:
```diff
@@ -18,7 +18,7 @@ use crate::{
 
 pub enum PinKeyboardMsg {
-    Confirmed,
+    Confirmed(Vec<u8, MAX_LENGTH>),
```
Since the `pin()` method exists I guess it was supposed to be used to obtain the pin, but haven't figured out how to get a reference to the layout object at the time the result is constructed.

jpochyla replied:
This is exactly the usecase the `ObjComponentMsg` trait (hope I remember the name right) is supposed to address.

jpochyla replied:
If you omit the `TryInto` impl for the msg type, you should be able to impl the above trait and get a ref to the component for the conversion. Then you can pull the PIN out of it and convert into Obj.

jpochyla replied:
If you omit the `TryInto` impl for the msg type, you should be able to impl the above trait and get a ref to the component for the conversion. Then you can pull the PIN out of it and convert into Obj.

mmilata replied:
When I implement the trait like https://github.com/trezor/trezor-firmware/blob/31fc2a4270b01cdbd4d8aeb33b34ba45065c9cb6/core/embed/rust/src/ui/model_tt/layout.rs#L57-L65
then I get:
```
error[E0119]: conflicting implementations of trait `ui::layout::obj::ComponentMsgObj` for type `ui::model_tt::component::keyboard::pin::PinKeyboard`
  --> src/ui/model_tt/layout.rs:57:1
   |
57 |   impl ComponentMsgObj for PinKeyboard
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `ui::model_tt::component::keyboard::pin::PinKeyboard`
   |
  ::: src/ui/layout/obj.rs:37:1
   |
37 | / impl<T> ComponentMsgObj for T
38 | | where
39 | |     T: Component,
40 | |     T::Msg: TryInto<Obj, Error = Error>,
...  |
44 | |     }
45 | | }
   | |_- first implementation here
```
What am I missing?

mmilata replied:
The `TryInto` implementation is not there anymore, perhaps the compiler is not smart enough to figure out that `T::Msg: TryInto<Obj, Error = Error>` is false?

mmilata replied:
As discussed, we may be hitting rust-lang/rust#20400. I rewrote all the `TryInto` conversions to explicit `ComponentMsgObj` impls, it's a bit ugly but uses less magic.

mmilata commented on core/embed/rust/src/ui/model_tt/component/keyboard/pin.rs:
```diff
@@ -145,7 +145,7 @@ impl Component for PinKeyboard {
     fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<Self::Msg> {
         if let Some(Clicked) = self.confirm_btn.event(ctx, event) {
-            return Some(PinKeyboardMsg::Confirmed);
+            return Some(PinKeyboardMsg::Confirmed(self.digits.clone()));
```
If we keep this, is there a way to zero out the pin being cloned in a way that doesn't get optimized out?

jpochyla replied:
See above, we can get rid of PIN being stored in the message altogether. Re. reliable data zeroing, there are some crates, let's investigate how they work in a nostd env.

jpochyla replied:
https://docs.rs/zeroize/latest/zeroize/ is such a crate.

mmilata replied:
Not storing the pin in the message anymore so zeroing isn't needed (yet). But good to know such crate exists.

matejcik replied:
maybe you should zero out the source buffer for the PIN though?

matejcik replied:
👍 
added the TODO

honestly, couldn't we just consume the layout object instead? `self` instead of `&mut self`
(probably not because something something reference in the Python-held object, but still something that seems semantically more correct)

mmilata replied:
#2552

mmilata commented on core/embed/rust/src/micropython/macros.rs:
```diff
@@ -164,3 +170,23 @@ macro_rules! obj_type {
+                },
+                // TODO: This is safe only if we pass in `Dict` with fixed `Map` (created by
+                // `Map::fixed()`, usually through `obj_map!`), because only then will
+                // MicroPython treat `globals` as immutable, and make the mutable cast safe.
```
Would it be possible/difficult/ugly to just call the `obj_dict!(obj_map!(...))` from here instead?

jpochyla replied:
Not difficult, just ugly, we'd need to copy the field matchers from the macro definitions into this one. But I guess we should. Or we can maybe make a newtype for the static dicts, but I don't like it too much.

mmilata replied:
Copied the field matchers in [681ac69](https://github.com/trezor/trezor-firmware/pull/2138/commits/681ac69b3f12cbb842e46f646cdd2047b6bcc00b).

matejcik replied:
the comment needs to be tweaked: "This is safe only because we are passing in a static dict with a fixed Map created by (...). That causes MicroPython to treat `globals` as immutable (...)"

mmilata replied:
744f4d6bfafbc0b7676ff3bdead8160af6441bba

matejcik commented:
we can extend `tools/build_mocks` relatively easily to also walk Rust files (maybe only `obj.rs` ?) and generate mocks in the same format as the `modtrezor*.h` files.

mmilata commented:
I've addressed the comments and extended `build_mocks` to also scan `.rs` file for definitions, please take a look. I think this shouldn't conflict with #2155 much so marking as ready for review.

mmilata commented:
Also rebased to master.
cc @grdddj

mmilata commented on core/embed/rust/src/micropython/macros.rs:
```diff
@@ -164,3 +170,36 @@ macro_rules! obj_type {
+            static DICT: ffi::mp_obj_dict_t = ffi::mp_obj_dict_t {
+                base: ffi::mp_obj_base_t {
+                    /// SAFETY: FIXME extern static
+                    type_: unsafe { &ffi::mp_type_dict },
```
@jpochyla can you advise if this is safe? Seems like a value that shouldn't be changing normally.

jpochyla replied:
Safe 👍

mmilata replied:
Thanks, updated comment in d128f680d9d2dd267407647e608b38f501a6d737.

matejcik reviewed (approved):
LGTM structure wise, leaving the details to @jpochyla

big :+1: on `build_mocks` handling

jpochyla commented on core/SConscript.unix:
```diff
@@ -668,9 +668,13 @@ protobuf_blobs = env.Command(
 
-RUST_PROFILE = 'dev' if ARGUMENTS.get('TREZOR_EMULATOR_DEBUGGABLE', 0) == '1' else 'release'
+if ARGUMENTS.get('TREZOR_EMULATOR_DEBUGGABLE', '0') == '1':
+    RUST_PROFILE = 'dev'
```
Why not "debug"?

mmilata replied:
With `debug` I'm getting:
```
$ make build_unix_debug
...
cd embed/rust; cargo build --profile debug --target-dir=../../build/unix/rust --no-default-features --features "model_tt"
error: profile name `debug` is reserved
To configure the default development profile, use the name `dev` as in [profile.dev]
```

According to https://doc.rust-lang.org/cargo/reference/profiles.html there's no default `debug` profile.

jpochyla replied:
Facepalm, sorry.

jpochyla commented on core/embed/rust/src/micropython/obj.rs:
```diff
@@ -289,6 +289,22 @@ impl TryFrom<&'static CStr> for Obj {
+        // SAFETY:
+        //  - Should work with any micropython objects.
+        // EXCEPTION: Will raise if allocation fails.
+        let obj = catch_exception(|| unsafe { ffi::mp_obj_new_tuple(2, [val.0, val.1].as_ptr()) })?;
```
Could we please put `[val.0, val.1]` into a variable?

mmilata replied:
Sure, fixed in f17120e0d330692e053d902c01f648021702cec4.

mmilata commented:
Noticed that `rustc` doesn't like `///` for the mocks:
```
warning: unused doc comment
   --> src/ui/model_tt/layout.rs:153:5
    |
153 |     /// CONFIRMED: object
    |     ^^^^^^^^^^^^^^^^^^^^^
154 |     Qstr::MP_QSTR_CONFIRMED => CONFIRMED.as_obj(),
    |     ----------------------- rustdoc does not generate documentation for expressions
    |
    = note: `#[warn(unused_doc_comments)]` on by default
    = help: use `//` for a plain comment
```

I've added `#[allow(unused_doc_comments)]` but let me know if using different prefix is preferable.

matejcik commented on core/embed/rust/src/ui/geometry.rs:
```diff
@@ -447,7 +447,7 @@ impl Grid {
     }
 
     pub fn cells(&self, cells: GridCellSpan) -> Rect {
```
I recall wanting to rename this to `cell_span` but apparently I didn't in my previous modifications? Please do that here.

matejcik replied:
actually let's not do that, I will be doing a subsequent PR so i will do that there separately

mmilata commented on core/embed/rust/src/ui/model_tt/layout.rs:
```diff
@@ -1,109 +1,124 @@
+    fn msg_try_into_obj(&self, msg: Self::Msg) -> Result<Obj, Error> {
+        match msg {
+            PinKeyboardMsg::Cancelled => Ok(CANCELLED.as_obj()),
+            PinKeyboardMsg::Confirmed => self.pin().try_into(),
```
good catch, fixed in [e75c848](https://github.com/trezor/trezor-firmware/pull/2138/commits/e75c848977b590b2a6a6d8a2878e3bbb39571c79)
